### PR TITLE
WhenRepeatedFilter - Filtering of identical LogEvents

### PR DESCRIPTION
**Pull request by snakefoot**
Useful if having a permanent error condition.

### CODE REVIEW

File: src/NLog/Filters/WhenRepeatedFilter.cs [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+            }
+            public string StringValue;
+            public readonly StringBuilder StringBuffer = new StringBuilder();
+            public int StringHashCode;
+            public LogLevel LogLevel;
+            public DateTime LastLogTime;
+            public int FilterCount;
+        }
+
+        private readonly List<FilterInfo> _mruFilter = new List<FilterInfo>();
```
Comment (304NotModified) [inline src/NLog/Filters/WhenRepeatedFilter.cs]: please move to top of the class

File: src/NLog/Filters/WhenRepeatedFilter.cs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+                    {
+                        value = Layout.Render(logEvent) ?? string.Empty;
+                        if (value.Length > MaxLength)
+                            value = value.Substring(0, MaxLength);
+                        hashCode = StringComparer.Ordinal.GetHashCode(value);
+                    }
+
+                    int optimalRemoveIndex = 0;
+
+                    for (int i = _mruFilter.Count - 1; i >= 0; i--)
```
Comment (304NotModified) [inline src/NLog/Filters/WhenRepeatedFilter.cs]: Please a small description why we move in opposite direction

File: src/NLog/Filters/WhenRepeatedFilter.cs [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        /// Can be used if <see cref="OptimizeBufferReuse"/> has been enabled.
+        /// </summary>
+        internal readonly ReusableBuilderCreator ReusableLayoutBuilder = new ReusableBuilderCreator();
+
+        public WhenRepeatedFilter()
+        {
+            MruCount = 50;
+            TimeoutSeconds = 10;
+            MaxLength = 1000;
+            OptimizeBufferReuse = true;
```
Comment (304NotModified) [inline src/NLog/Filters/WhenRepeatedFilter.cs]: Please add the [DefaultValue] attributes on the properties of those

File: src/NLog/Filters/WhenRepeatedFilter.cs [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    using System;
+    using System.Collections.Generic;
+    using System.Text;
+    using NLog.Internal;
+
+    /// <summary>
+    /// Matches when the result of the calculated layout has been repeated a moment ago
+    /// </summary>
+    [Filter("whenRepeated")]
+    class WhenRepeatedFilter : LayoutBasedFilter
```
Comment (304NotModified) [inline src/NLog/Filters/WhenRepeatedFilter.cs]: please add the modifier

File: src/NLog/Filters/WhenRepeatedFilter.cs [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+                    string value = string.Empty;
+                    int hashCode = 0;
+                    if (OptimizeBufferReuse)
+                    {
+                        Layout.RenderAppendBuilder(logEvent, targetBuilder.Result, false);
+                        hashCode = Math.Min(targetBuilder.Result.Length, MaxLength);
+                    }
+                    else
+                    {
+                        value = Layout.Render(logEvent) ?? string.Empty;
```
Comment (304NotModified) [inline src/NLog/Filters/WhenRepeatedFilter.cs]: Afaik Render never returns `null`, but nothing wrong to keep it like this

File: src/NLog/Filters/WhenRepeatedFilter.cs [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+
+        public WhenRepeatedFilter()
+        {
+            MruCount = 50;
+            TimeoutSeconds = 10;
+            MaxLength = 1000;
+            OptimizeBufferReuse = true;
+        }
+
+        protected override FilterResult Check(LogEventInfo logEvent)
```
Comment (304NotModified) [inline src/NLog/Filters/WhenRepeatedFilter.cs]: Please add some XML docs (or `<inheritdoc />`

File: src/NLog/Filters/WhenRepeatedFilter.cs [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+        {
+            lock (_mruFilter)
+            {
+                using (var targetBuilder = OptimizeBufferReuse ? ReusableLayoutBuilder.Allocate() : ReusableLayoutBuilder.None)
+                {
+                    string value = string.Empty;
+                    int hashCode = 0;
```
Comment (304NotModified) [inline src/NLog/Filters/WhenRepeatedFilter.cs]: the default of hashcode could be removed

File: src/NLog/Filters/WhenRepeatedFilter.cs [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+                    }
+
+                    int optimalRemoveIndex = 0;
+
+                    for (int i = _mruFilter.Count - 1; i >= 0; i--)
+                    {
+                        var filterInfo = _mruFilter[i];
+
+                        if (filterInfo.FilterCount == 0 || (logEvent.TimeStamp - filterInfo.LastLogTime).TotalSeconds > TimeoutSeconds)
+                            optimalRemoveIndex = i;
```
Comment (304NotModified) [inline src/NLog/Filters/WhenRepeatedFilter.cs]: what are we doing here?

File: src/NLog/Filters/WhenRepeatedFilter.cs [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+                                continue;
+                        }
+                        else
+                        {
+                            if (!string.Equals(filterInfo.StringValue, value, StringComparison.Ordinal))
+                                continue;
+                        }
+
+                        FilterResult filterResult = FilterResult.Neutral;
+                        if ((logEvent.TimeStamp - filterInfo.LastLogTime).TotalSeconds > TimeoutSeconds || logEvent.Level.Ordinal > filterInfo.LogLevel.Ordinal)
```
Comment (304NotModified) [inline src/NLog/Filters/WhenRepeatedFilter.cs]: this is a duplicate check? (`(logEvent.TimeStamp - filterInfo.LastLogTime).TotalSeconds > TimeoutSeconds`)

File: src/NLog/Filters/WhenRepeatedFilter.cs [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+                            continue;
+
+                        if (OptimizeBufferReuse)
+                        {
+                            bool stringEquals = true;
+                            int length = Math.Min(targetBuilder.Result.Length, MaxLength);
+                            if (length != filterInfo.StringBuffer.Length)
+                                continue;
+
+                            for (int x = 0; x < length; ++x)
```
Comment (304NotModified) [inline src/NLog/Filters/WhenRepeatedFilter.cs]: please think of another name for `x`

File: src/NLog/Filters/WhenRepeatedFilter.cs [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+                            int filterCount = filterInfo.FilterCount;
+                            filterInfo.Refresh(logEvent.Level, logEvent.TimeStamp, 0);
+                            if (filterCount > 0 && !string.IsNullOrEmpty(FilterCountPropertyName))
+                            {
+                                object otherFilterCount;
+                                if (!logEvent.Properties.TryGetValue(FilterCountPropertyName, out otherFilterCount))
+                                {
+                                    logEvent.Properties[FilterCountPropertyName] = filterCount;
+                                }
+                                else if (otherFilterCount is int)
```
Comment (304NotModified) [inline src/NLog/Filters/WhenRepeatedFilter.cs]: we have to move to c# 7 soon ;) ` (otherFilterCount is int otherFilterCount)` - or something like that

File: src/NLog/Filters/WhenRepeatedFilter.cs [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+                    reusableObject.Update(value, hashCode, targetBuilder.Result, MaxLength);
+                    reusableObject.Refresh(logEvent.Level, logEvent.TimeStamp, 0);
+                    _mruFilter.Add(reusableObject);
+                }
+            }
+
+            return FilterResult.Neutral;
+        }
+
+        class FilterInfo
```
Comment (304NotModified) [inline src/NLog/Filters/WhenRepeatedFilter.cs]: please add modifier and docs.

File: src/NLog/Filters/WhenRepeatedFilter.cs [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+                LastLogTime = lastLogTime;
+                FilterCount = filterCount;
+                LogLevel = logLevel;
+            }
+            public string StringValue;
+            public readonly StringBuilder StringBuffer = new StringBuilder();
+            public int StringHashCode;
+            public LogLevel LogLevel;
+            public DateTime LastLogTime;
+            public int FilterCount;
```
Comment (304NotModified) [inline src/NLog/Filters/WhenRepeatedFilter.cs]: I think most of them could be `private`?

**Comment by snakefoot** [comment]
@304NotModified Have changed it from a MostRecentlyUsed (mru)-list to a Dictionary. Also improved performance by using StringBuilder.Equals, though it requires some special capacity-conditions.

**Comment by 304NotModified** [comment]
great feature!!

**Comment by 304NotModified** [comment]
needs some docs I guess?

**Comment by snakefoot** [comment]
@304NotModified Have added some docs:

https://github.com/NLog/NLog/wiki/WhenRepeated-Filter

**Pull request merged**